File: voice_loop/console/panels/voices.py
"""面板：角色与声线。

能做的事：看每个角色的人设要点、唤醒词、参考音频与专属模型目录；
**切过去**（服务里真的换声线）；**试听一句**（用服务已加载的 TTS 念出来）。

★真正出声的一定是服务★：控制台自己加载 ZipVoice 要好几百 MB 到几 GB，
而且两个进程抢同一张声卡没好处。所以这里全部走信箱（``voice_loop/control.py``），
服务没在跑时如实报错，不假装成功。
"""

from __future__ import annotations

from ..registry import Panel

PANEL = Panel(
    id="voices",
    title="角色 / 声线",
    order=50,
    hint="看角色、切声线、试听（需要服务在跑）",
    needs_service=True,
)


def register(app, ctx) -> None:
    from fastapi import Body, HTTPException  # noqa: PLC0415

    def character_rows() -> list[dict]:
        registry = ctx.characters()
        default_id = str(ctx.settings.persona.default or "")
        resolved = registry.get(default_id) or registry.default()
        resolved_id = getattr(resolved, "id", "") or default_id
        rows = []
        for char in registry.all():
            ref = str(getattr(char, "voice_ref", "") or "")
            ref_path = ctx.settings.resolve(ref) if ref else None
            ref_files = []
            if ref_path is not None and ref_path.parent.is_dir():
                ref_files = sorted(
                    f"{p.relative_to(ctx.root).as_posix()}"
                    for p in ref_path.parent.glob("*.wav")
                )
            rows.append({
                "id": char.id,
                "name": char.name,
                "title": char.title,
                "background": (char.background or "")[:400],
                "enabled": bool(char.enabled),
                "is_default": bool(char.default) or char.id == resolved_id,
                "user_title": char.user_title,
                "wake_words": list(char.wake_words or []),
                "aliases": {k: list(v or []) for k, v in (char.aliases or {}).items()},
                "ack": char.ack or "",
                "style": list(char.style or []),
                "voice": char.voice or "",
                "voice_ref": ref,
                "voice_ref_exists": bool(ref_path and ref_path.exists()),
                "voice_ref_text": char.voice_ref_text or "",
                "voice_model": char.voice_model or "",
                "voice_model_exists": bool(
                    char.voice_model and ctx.settings.resolve(char.voice_model).is_dir()
                ),
                "voice_dir": char.voice_dir or "",
                "ref_candidates": ref_files[:40],
                "lines": [{"scene": ln.get("scene", ""), "text": ln.get("text", "")}
                          for ln in (char.lines or [])][:8],
                "notes": char.notes or "",
            })
        return rows

    @app.get("/api/voices")
    def api_list():
        settings = ctx.settings
        try:
            chars = character_rows()
            default_name = next((c["name"] for c in chars if c["is_default"]), "")
            error = ""
        except Exception as exc:  # noqa: BLE001 - 角色文件坏了也要能打开这一页
            chars, default_name, error = [], "", f"{type(exc).__name__}: {exc}"
        return {
            "characters": chars,
            "error": error,
            "default": settings.persona.default,
            "resolved_default": default_name or settings.persona.default,
            "persona_file": settings.persona.file,
            "tts": {
                "backend": settings.tts.backend,
                "clone_dir": settings.tts.clone_dir,
                "clone_audio": settings.tts.clone_audio,
                "clone_steps": settings.tts.clone_steps,
                "model": settings.tts.model,
            },
            "mailbox": ctx.channel.status(),
        }

    @app.post("/api/voices/switch")
    def api_switch(payload: dict = Body(default={})):
        """切声线：交给服务做（它持有引擎），并等回执确认到底切没切。"""
        cid = str((payload or {}).get("id") or "").strip()
        if not cid:
            raise HTTPException(status_code=400, detail="缺少 id")
        reply = ctx.call_service("character", timeout=float((payload or {}).get("timeout") or 20.0), id=cid)
        return {
            "ok": reply.ok,
            "message": reply.text if reply.ok else reply.error,
            "data": reply.data,
            "seconds": round(reply.seconds, 2),
        }

    @app.post("/api/voices/audition")
    def api_audition(payload: dict = Body(default={})):
        """试听：先切（可选）再念一句。两步都走信箱，每一步都要回执。"""
        body = payload or {}
        cid = str(body.get("id") or "").strip()
        text = str(body.get("text") or "").strip() or "你好呀，我是本地语音助手。"
        switched = None
        if cid and body.get("switch", True):
            reply = ctx.call_service("character", timeout=30.0, id=cid)
            switched = {"ok": reply.ok, "message": reply.text if reply.ok else reply.error}
            if not reply.ok:
                return {"ok": False, "error": "切换失败：" + reply.error, "switch": switched}
        # 切声线后 TTS 要卸载重载 + 念一句，给足时间（实测首次 10 秒量级）
        say = ctx.call_service("say", timeout=float(body.get("timeout") or 60.0), text=text)
        return {
            "ok": say.ok,
            "error": say.error,
            "text": say.text,
            "seconds": round(say.seconds, 2),
            "switch": switched,
            "data": say.data,
        }

    @app.get("/api/voices/lines")
    def api_lines(cid: str = ""):
        """拿某个角色的示例台词，方便一键试听（听的要是她真会说的话）。"""
        for row in character_rows():
            if row["id"] == cid:
                return {"id": cid, "lines": row["lines"], "wake_words": row["wake_words"]}
        raise HTTPException(status_code=404, detail=f"没有角色 {cid!r}")

File: voice_loop/console/static/panels/voices.js
/* 角色 / 声线面板：看人设与参考音、切声线、试听（切过去再念一句）。 */
(function () {
  let handleEvent = null;

  window.Console.register("voices", {
    render(root, ctx) {
      const { h, card, table, secs } = UI;
      const text = h("input", { value: "你好呀，我是本地语音助手。", placeholder: "要念的话（可留空用默认）" });
      const listBox = h("div", {});

      root.appendChild(card("试听（声音由服务放出来）", [
        h("div", { class: "row" }, [
          h("div", { class: "grow" }, text),
          h("button", {
            class: "btn", text: "用当前声线念这句",
            onclick: async (e) => {
              e.target.disabled = true;
              const r = await ctx.api.safe(() => ctx.api.post("/api/voices/audition", { text: text.value, switch: false }));
              e.target.disabled = false;
              if (r) ctx.toast(r.ok ? "已念出（" + secs(r.seconds) + "）" : "没念成：" + r.error, r.ok ? "ok" : "err");
            },
          }),
        ]),
        h("p", { class: "muted", text: "点下面某个卡片的「切过去并试听」= 先切声线再念这句（两步都走信箱，服务最多 5 秒内响应）。服务没启动会失败。" }),
      ]));
      root.appendChild(listBox);

      async function audition(id) {
        const r = await ctx.api.safe(() => ctx.api.post("/api/voices/audition", { id, text: text.value }));
        if (r) ctx.toast(r.ok ? "已切到该角色并念出（" + secs(r.seconds) + "）" : "没成：" + r.error, r.ok ? "ok" : "err");
        refresh();
      }

      async function switchTo(id) {
        const r = await ctx.api.safe(() => ctx.api.post("/api/voices/switch", { id }));
        if (r) ctx.toast(r.ok ? "已切到 " + r.message : "没切成：" + r.message, r.ok ? "ok" : "err");
        refresh();
      }

      function render(d) {
        listBox.innerHTML = "";
        if (d.error) {
          listBox.appendChild(card("角色文件有问题", UI.line(d.error, "mono")));
        }
        const chars = d.characters || [];
        listBox.appendChild(card("角色（默认 " + (d.resolved_default || d.default || "—") + " · 人格文件 " + d.persona_file + "）",
          h("div", { class: "grid" }, chars.map((c) => card(c.name + (c.is_default ? " ·默认" : ""), [
            h("p", { class: "muted", text: c.title || "" }),
            h("p", { class: "muted", text: "唤醒词：" + ((c.wake_words || []).join("、") || "—") }),
            h("p", { class: "muted", text: "称呼你：「" + (c.user_title || "你") + "」" + (c.ack ? " · 应答语：" + c.ack : "") }),
            c.style && c.style.length ? h("p", { class: "muted", text: "风格：" + c.style.join(" / ") }) : null,
            h("p", { class: "mono", text: "参考音：" + (c.voice_ref || "（用全局）") + (c.voice_ref ? (c.voice_ref_exists ? " ✓" : " ✗ 文件不在") : "") }),
            c.voice_model ? h("p", { class: "mono", text: "专属模型：" + c.voice_model + (c.voice_model_exists ? " ✓" : " ✗ 目录不在") }) : null,
            c.ref_candidates && c.ref_candidates.length
              ? h("details", {}, [
                  h("summary", { class: "muted", text: "素材库里还有 " + c.ref_candidates.length + " 个参考候选" }),
                  h("div", { class: "mono", style: "font-size:11px" }, c.ref_candidates.map((p) => h("div", { text: p }))),
                ])
              : null,
            c.lines && c.lines.length
              ? h("details", {}, [
                  h("summary", { class: "muted", text: "示例台词 " + c.lines.length + " 条" }),
                  h("div", {}, c.lines.map((ln) => h("div", { class: "muted", text: "· " + (ln.text || "") }))),
                ])
              : null,
            h("div", { class: "row" }, [
              h("button", { class: "btn small primary", text: "切过去并试听", onclick: () => audition(c.id) }),
              h("button", { class: "btn small", text: "只切过去", onclick: () => switchTo(c.id) }),
            ]),
          ])))));

        listBox.appendChild(card("全局声线配置（只读；要改去 config.toml 或角色 json）", table(
          [{ key: "k", label: "项", cls: "num" }, { key: "v", label: "值" }],
          [
            { k: "TTS 后端", v: d.tts.backend },
            { k: "克隆模型目录", v: d.tts.clone_dir },
            { k: "克隆参考音（全局）", v: d.tts.clone_audio },
            { k: "采样步数", v: d.tts.clone_steps },
            { k: "Piper 模型", v: d.tts.model },
            { k: "信箱待处理", v: d.mailbox.pending + " 条" },
          ]
        )));
      }

      function refresh() {
        ctx.api.get("/api/voices")
          .then(render)
          .catch((err) => ctx.toast("读取角色失败：" + err.message, "err"));
      }
      refresh();

      // 语音里喊了角色名切换 → 界面跟着更新
      handleEvent = (ev) => {
        if (ev.kind === "log" && ev.data && /\[角色\]/.test(ev.data.line || "")) {
          clearTimeout(refresh._t);
          refresh._t = setTimeout(refresh, 1200);
        }
      };
    },

    onEvent(ev, ctx) { if (handleEvent) handleEvent(ev, ctx); },
    onLeave() { handleEvent = null; },
  });
})();
